### PR TITLE
Added optional endowments param to executeSnap

### PR DESCRIPTION
**Pull request by shanejonas**
This adds an optional `endowments` parameter to `executeSnap` to allow access to globals like `fetch` per snap.

##### Example with `fetch` endowment
image

image

##### Example with no endowments:
image

### CODE REVIEW

File: openrpc.json [hunk 1]
```diff
@@ -37,6 +37,19 @@
             "title": "SourceCode",
             "type": "string"
           }
+        },
+        {
+          "name": "endowments",
+          "description": "a list off endowments to provide to SES",
```
Comment (shanejonas) [inline openrpc.json:43]:
```suggestion
          "description": "a list of endowments to provide to SES",
```

File: src/controller.ts [hunk 2]
```diff
@@ -215,6 +223,12 @@ class Controller {
       XMLHttpRequest,
     };
 
+    if (_endowments && _endowments.length > 0) {
+      _endowments.forEach((_endowment) => {
+        endowments[_endowment] = (window as any)[_endowment].bind(window);
+      });
+
+
```
Comment (shanejonas) [inline src/controller.ts:231]:
```suggestion
```

File: src/controller.ts [hunk 3]
```diff
@@ -200,13 +208,13 @@ class Controller {
...
 
-    const endowments = {
+    const endowments: Record<string, any> = {
       BigInt,
       Buffer,
       console, // Adding raw console for now
       crypto: window.crypto,
       Date,
-      fetch: window.fetch.bind(window),
+      // fetch: window.fetch.bind(window),
```
Comment (shanejonas) [inline src/controller.ts:217]:
```suggestion
```

File: src/controller.ts [hunk 4]
```diff
@@ -215,6 +223,11 @@ class Controller {
       XMLHttpRequest,
```
Comment (shanejonas) [inline src/controller.ts:223]:
```suggestion
```

File: openrpc.json [hunk 5]
```diff
@@ -37,6 +37,19 @@
             "title": "SourceCode",
             "type": "string"
           }
+        },
+        {
+          "name": "endowments",
+          "description": "a list of endowments to provide to SES",
```
Comment (rekmarks) [inline openrpc.json:43]:
```suggestion
          "description": "A list of endowments to provide to SES",
```

File: src/controller.ts [hunk 6]
```diff
@@ -215,6 +223,11 @@ class Controller {
...
     };
 
+    if (_endowments && _endowments.length > 0) {
+      _endowments.forEach((_endowment) => {
+        endowments[_endowment] = (window as any)[_endowment].bind(window);
+      });
```
Comment (rekmarks) [inline src/controller.ts:229]: Some defensive improvements:
```suggestion
    if (_endowments && _endowments.length > 0) {
      _endowments.forEach((_endowment) => {
        if (!(_endowment in window)) {
          throw new Error(`Unknown endowment: "${_endowment}".`);
        }

        
        endowments[_endowment] = typeof globalValue === 'function'
          ? globalValue.bind(window)
          : globalValue;
      });
```

File: src/controller.ts [hunk 7]
```diff
@@ -200,13 +208,13 @@ class Controller {
...
+    const endowments: Record<string, any> = {
       BigInt,
       Buffer,
       console, // Adding raw console for now
       crypto: window.crypto,
       Date,
-      fetch: window.fetch.bind(window),
+      // fetch: window.fetch.bind(window),
       Math, // Math.random is considered unsafe, but we need it
       setTimeout,
```
Comment (rekmarks) [inline src/controller.ts:219]: Note for follow-up: `setTimeout` is a sandbox escape. We have to fix it.

**Review by rekmarks — CHANGES_REQUESTED** [review]

File: src/controller.ts (hunk 4 above)
Comment (rekmarks) [reply src/controller.ts:223]: `WebSocket` also needs to go, unless we decide to include it.

File: src/controller.ts [hunk 8]
```diff
@@ -12,7 +12,11 @@ import EEOpenRPCDocument from '../openrpc.json';
 import { STREAM_NAMES } from './enums';
 
 import { IframeExecutionEnvironmentMethodMapping, methods } from './methods';
-import { JSONRPCRequest } from './__GENERATED_TYPES__';
+import {
+  Endowments,
+  JSONRPCRequest,
+  StringDoaGddGA,
```
Comment (shanejonas) [inline src/controller.ts:18]:
```suggestion
```

**Review by rekmarks — APPROVED** [review]
LGTM!

**Pull request merged**